File: demo/business-practice/movie-rating-synthesis/src/main/scala/yore/dt/SecondarySortingKey.scala
package yore.dt

/**
  *
  * Created by yore on 2019/2/22 13:51
  */
class SecondarySortKey(val first: Double, val second: Double)
  extends Ordered[SecondarySortKey] with Serializable {

  override def compare(that: SecondarySortKey): Int = {
    if(this.first - that.first !=0){
      (this.first - that.first).toInt
    }else{
      if(this.second -that.second >0 ){
        Math.ceil(this.second - that.second).toInt
      }else if(this.second - that.second <0){
        Math.floor(this.second - that.second).toInt
      }else{
        (this.second - that.second).toInt
      }
    }
  }
}

/*class SecondarySortKey(val first : Double, val second : Double) extends Ordered[SecondarySortKey] with Serializable {

  override def compare(other : SecondarySortKey): Int = {
    //先判断第一个排序字段是否相等，如果不相等，就直接排序
    if(this.first - other.first != 0){
      (this.first - other.first).toInt
    }else{
      // 如果第一个字段相等，则比较第二个字段，若实现多次排序则以此判断
      if(this.second - other.second >0){
        Math.ceil(this.second - other.second).toInt
      }else if(this.second - other.second < 0 ){
        Math.floor(this.second - other.second).toInt
      }else{
        (this.second - other.second).toInt
      }
    }

  }
}*/

File: case/business-trilogy/movie-rating-synthesis/src/main/java/yore/secondary/SecondarySortingKey.java
package yore.secondary;

import scala.Serializable;
import scala.math.Ordered;

/**
 * 自定义Key值类
 *
 * Created by yore on 2019/2/22 12:30
 */
public class SecondarySortingKey implements Ordered<SecondarySortingKey>, Serializable {
    private int first;
    private int second;

    public SecondarySortingKey(int first, int second) {
        this.first = first;
        this.second = second;
    }

    @Override
    public int compare(SecondarySortingKey that) {
        // 如果第一个值不相等，直接比较第一个值即可，如果第一个相等，则比较第二个值
        if(this.first - that.getFirst() != 0){
            return this.first - that.getFirst();
        }else{
            return this.second - that.getSecond();
        }
    }

    @Override
    public boolean $less(SecondarySortingKey that) {
        if(this.first < that.getFirst()){
            return true;
        }else if(that.first == that.getFirst() && that.second < that.getSecond()){
            return true;
        }
        return false;
    }

    @Override
    public boolean $less$eq(SecondarySortingKey that) {
        if(SecondarySortingKey.this.$less(that)){
            return true;
        }else if(that.first == that.getFirst() && this.second == that.getSecond()){
            return true;
        }
        return false;
    }

    @Override
    public boolean $greater(SecondarySortingKey that) {
        if(this.first > that.getFirst()){
            return true;
        }else if(that.first == that.getFirst() && this.second > that.getSecond()){
            return true;
        }
        return false;
    }



    @Override
    public boolean $greater$eq(SecondarySortingKey that) {
        if(SecondarySortingKey.this.$greater(that)){
            return true;
        }else if(that.first == that.getFirst() && this.second == that.getSecond()){
            return true;
        }
        return false;
    }

    @Override
    public int compareTo(SecondarySortingKey that) {
        if(this.first - that.getFirst() !=0){
            return this.first - that.getFirst();
        }else{
            return this.second - that.getSecond();
        }
    }


    public int getFirst() {
        return first;
    }
    public void setFirst(int first) {
        this.first = first;
    }

    public int getSecond() {
        return second;
    }
    public void setSecond(int second) {
        this.second = second;
    }

    @Override
    public String toString() {
        return super.toString();
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        SecondarySortingKey that = (SecondarySortingKey) o;
        return first == that.first &&
                second == that.second;
    }

    @Override
    public int hashCode() {
        int result = first;
        result = 31 * result + second;
        return result;
//        return Objects.hash(first, second);
    }
}
